### PR TITLE
orders and revenue page

### PR DESCRIPTION
**Pull request by fivetran-reneeli**
Streamlit script and configs for the orders and revenue page

### CODE REVIEW

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: This code will always always show new customers of 0 since it is only counting a new customer if they were created on the exact same day as the order being processed.

Instead I would recommend the following changes. This ensures we are counting a new customer in the period. If the customer created date is between the min and max created_at timestamps, then we can count them as a new customer in the period.
```suggestion
    min_created_at = data['created_at'].min()
    max_created_at = data['created_at'].max()
    new_customers = data[(data['customer_created_at'] >= min_created_at) & (data['customer_created_at'] <= max_created_at)].shape[0]
```

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]:
```suggestion
        st.metric(label="**Number of Customers in Period**", value=number_of_customers)
```

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: Let's move this above the customer table.

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: This chart is a bit hard to read on the white background. Here are a few suggestions to apply to make things pop a bit more and also to make the hover components a bit less crowded.

```suggestion
# Sort by total_amount in descending order
location_performance = location_performance.sort_values(by='total_amount', ascending=False)

# Format total_amount for better readability
location_performance['total_amount_formatted'] = location_performance['total_amount'].apply(lambda x: f"${x:,.0f}")

# Define a custom color scale that starts with darker shades of green
custom_color_scale = [
    (0.0, "rgb(198, 219, 239)"),
    (0.2, "rgb(158, 202, 225)"),
    (0.4, "rgb(107, 174, 214)"),
    (0.6, "rgb(66, 146, 198)"),
    (0.8, "rgb(33, 113, 181)"),
    (1.0, "rgb(8, 69, 148)")
]

# Create the Plotly Express choropleth map
fig = px.choropleth(
    location_performance,
    locations='customer_country',
    locationmode='country names',
    color='total_amount',
    color_continuous_scale=custom_color_scale,
    title='Revenue by Country',
    labels={'customer_country': 'Country', 'total_amount_formatted': 'Total Revenue'}
)
```

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: This chart really isn't showing us anything interesting. Since we removed the Successful payments over time KPI I would recommend we change this chart to show something different. Any ideas? I would just request we use this chart to show a graph that tells a story with this data.

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: I noticed you used a line graph for this visualization. However this tells a misleading story. When looking at the chart the March 2024 month looks like it is nearly at 0. However, it is actually at 680k. I find it misleading that our 0 gridpoint is actually 680k. Can we change this to be a bar chart and show the total revenue from 0 up each month.

Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: Same comment for the Total number of orders and new customers charts

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: This is great, I think it would be interesting to also include total discounts. Would you be able to add that in as well?

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli great work on this PR! I have a few suggestions, recommendations, and requests to address before this is good to approve. Let me know if you want to sync on any of my notes. Thanks!

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py:29]: The second divider is no longer necessary now that the example dataframe is removed.
```suggestion
```

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Yeah the logic I used was to count the unique payment_ids whenever status = completed. However, as in the sample data's case, the payment_id only exists if the payment is successful I think, which is why it becomes 100% always. But it may be the same case in some finance platforms based on the logic we wrote. So I agree on taking it out. 

Maybe we can swap to Average Order Size?

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Can you show me where you saw it looking like 0? this is what I see... but happy to switch to bar chart and make sure the axis starts from 0.

image

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: added!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: switched!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py:29]: moved to below the KPI section

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Ah good catch!

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: @fivetran-reneeli it doesn't explicitly start from 0, but when looking at the chart it looks likeJuly is all the way at the bottom (visual representation of 0). If we make this a bar chart then we can highlight the entirety of the month and not just the change month over month.

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: @fivetran-reneeli I like where you are thinking. Maybe we do something to showcase the products being sold? I am thinking of a chart that helps a customer see which products are selling best over time. Or something along those lines. Thoughts?

Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: I see-- I think we can either do 
1. stacked bar chart of revenue over time that's broken up by product.
2. horizontal bar chart representing products, & the cumulative revenue for the given time period

I am leaning towards the latter since 1 is just pretty much the same as Revenue over Time which we have.

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: updated-- let me know what you think!

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: I agree - Let's go with the latter!

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: No need for a divider here.
```suggestion
```

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: Let's make the update here that was discussed in the last review to focus on product revenue

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: Thanks for making these bar charts as I believe that tells a more comprehensive story. However, these charts now look very similar and leads to making the charts seem not that exciting. 

What are your thoughts on making these combo charts and adding a line to each of these which will show the average revenue and average order by month?

Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: Additionally, what are your thoughts on adding permanent labels to make this easier to read right away? Similar to the following:

image

This is not required, but could be a nice feature to add to the accessibility of the charts. [Here](https://github.com/fivetran-streamlit/standardized_billing_model/blob/fa8af19e0c720792af04c06ca119acc522d2b5fd/pages/billing_report.py#L166-L177) is some code that shows how I added those text labels to the chart.

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: After reviewing @fivetran-catfritz PR I really liked the inclusion of the delta to these KPIs which helps tell the narrative of how the fake data is doing compared to the previous X period. 

https://github.com/fivetran/streamlit_fivetran_billing_model/blob/91011fcd634c53d16d562d7faafeb14fcec1546a/pages/2_subscriptions_report.py#L57-L61

Would you be able to add a delta to each of these metrics? I believe a Year over Year or Quarter over Quarter delta comparison would suffice here.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli thanks for the quick turnaround! I have a few final comments following your changes and also following @fivetran-catfritz PR review that I would like to apply to your PR. 

Let me know if you would like to sync on any of my comments, or the new product revenue chart as I know that is a net new chart to replace the successful subscriptions.

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Updated1

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: removed

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: updated!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Added YoYs!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: updated

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: Update looks good, thanks!

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: @fivetran-reneeli this looks good, but the colors are a bit confusing. What information are the different colors providing for each product? If we are doing a vertical bar chart in this way I imagine the colors could all be the same, right? 

Maybe doing a gradient would be more appropriate?

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: @fivetran-reneeli the data labels look great and make this chart way easier to read at first view. Thanks for that update!

> What are your thoughts on making these combo charts and adding a line to each of these which will show the average revenue and average order by month?

I was thinking we keep the two separate charts (one for revenue and one for orders), but then add a line component to each which highlights the rolling overage for month over month. So the Revenue chart would have a line graph included in it which would show the average revenue per month and how it changes month over month. 

I noticed you took a different approach and combined revenue and orders into one chart. The revenue being the bar chart and the orders the line. I like this approach as it helps to reduce repetitive information as both charts were essentially telling the same story.

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: What would the LOE be to adjust this to QoQ? Just to fit into the limitation of our fake data not always having a full year to look back?

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: The green color looks a bit off when compared to the other blue themed charts of the page. Would you be able to switch this to blue for consistency.

Additionally, I feel New Customers All Time is more information than necessary. Can we limit this to New Customers within the date filtered time frame?

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [inline pages/1_orders_and_revenue.py]: How do you feel the orientation of the chart looks when viewed alongside the other charts on the page? I feel like the vertical orientation looks a bit off and throws off the flow of the page. What are your thoughts? Should we consider flipping to horizontal for consistency?

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-reneeli great work on these updates! I have responded to a few comments above and also have a few last comments before approval. Let me know if you have any questions. Thanks!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Ah yes I had that in initially, but I was thinking it would look confusing to work with the date filter. Like if we picked June-July 2023, how does that get translated to a quarter? Same with if we did Jan-July 2023, how would a quarter get broken up? Whereas the YoY is more straightforward.

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: It's very minor but I associate a horizontal chart with time series. But let me try that and lmk what you think

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Oh I see, I interpreted as combining revenue and orders into one to reduce the duplicity of the 2 charts. But glad you like it!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: Sure, updated

File: pages/1_orders_and_revenue.py
Comment (fivetran-joemarkiewicz) [reply pages/1_orders_and_revenue.py]: There should be a native dataframe modification operation to group by quarter. See below as an example. 

https://github.com/fivetran/streamlit_fivetran_billing_model/blob/471ea1aceb83069a2146f1bbf3f4435660104fc2/pages/2_subscriptions_report.py#L58

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: updated the filter!

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: switched back to vertical

File: pages/1_orders_and_revenue.py
Comment (fivetran-reneeli) [reply pages/1_orders_and_revenue.py]: updated to include QoQ!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM! @fivetran-reneeli great work here! This is good to be merged into `main`.

**Pull request merged**
